### DISCUSSION TITLE
Utilities for creation & injection of AST nodes

### DISCUSSION THREAD

**Discussion by cdaringe** [syntax-tree/unist#38] [Q&A] [answered]
# Problem

Many utilities are listed for search, filter, parse, etc, but I see none for creation and node injection.

# Discussion

What approaches do you use? Plain old javascript?

I was hoping for something like `insertAfter(root, target, u(...myNewNode))`, akin to DOM APIs

I did consult the documentation site and various [lists of utilities](https://github.com/syntax-tree/mdast#list-of-utilities) pre post :)

**Comment by ChristianMurphy** [comment, marked as answer]
Personally I tend to use [unist-util-visit](https://github.com/syntax-tree/unist-util-visit) and [unist-util-visit-parents](https://github.com/syntax-tree/unist-util-visit-parents) the most, in both cases, mutating a `node` will modify the original AST.
For particularly complex node access, I may use [unist-util-select](https://github.com/syntax-tree/unist-util-select) to be able to leverage CSS-like selectors.

Another approach, specifically for building nodes is to use JSX syntax.
Which is built into [xastscript](https://github.com/syntax-tree/xastscript#jsx) and [hastscript (in a yet to be released version)](https://github.com/syntax-tree/hastscript/pull/15). it can also be applied to other AST builders.

> I was hoping for something like `insertAfter(root, target, u(...myNewNode))`, akin to DOM APIs

More DOM style APIs could be good contributions to unist/syntax-tree! Are you interested in creating some?
Some related ideas could include building out more React style APIs, expanding JSX support and maybe creating a `ref` like concept/util as another way to modify nodes.
Or building a jQuery-like API (which while it gets a lot of hate, inspired and helped bring modern APIs like `querySelector`)

**Reply by cdaringe** [reply]
I will never hate on jQuery :) It was the bee's knees! Until it wasn't!

> Personally I tend to...

I needed to parse, `<do async work>`, then re-find and append. I found that it worked fine to just mutate and print w/ the md AST printer.

> More DOM style APIs could be good contributions to unist/syntax-tree! Are you interested in creating some?

Maybe so! I've still evaluating this toolkit. I don't want to add libs for the sake of adding libs :). But if I continue to use it and am compelled, I'll be glad to send patches

**Reply by ChristianMurphy** [reply]
Async visitors are also supported, there is example how, here: https://github.com/syntax-tree/unist-util-visit-parents/issues/8
